prov/util/src/util_coll.c

```diff
@@ -1003,17 +1021,17 @@ int ofi_join_collective(struct fid_ep *ep, fi_addr_t coll_addr,
 		coll_mc = (struct util_coll_mc*) ((uintptr_t) coll_addr);
 	}
 
-	ret = util_coll_mc_alloc(&new_coll_mc);
-	if (ret)
-		return ret;
+	new_coll_mc = av_set->coll_mc;
+	if (!new_coll_mc) {
+		ret = util_coll_mc_alloc(&new_coll_mc);
+		if (ret)
+			goto err1;
+	}
 
 	util_ep = container_of(ep, struct util_ep, ep_fid);
 
 	// set up the new mc for future collectives
-	new_coll_mc->mc_fid.fid.fclass = FI_CLASS_MC;
 	new_coll_mc->mc_fid.fid.context = context;
-	new_coll_mc->mc_fid.fid.ops = &util_coll_fi_ops;
-	new_coll_mc->mc_fid.fi_addr = (uintptr_t) new_coll_mc;
 	new_coll_mc->av_set = av_set;
 	new_coll_mc->ep = ep;
```

Comment:
The above 2 lines can move into the coll_mc_alloc call.  You can consider moving setting the context into there as well.